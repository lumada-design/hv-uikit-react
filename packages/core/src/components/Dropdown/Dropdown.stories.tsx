import { useEffect, useMemo, useState } from "react";
import { css } from "@emotion/css";
import { fireEvent, screen, waitFor } from "@storybook/testing-library";
import { DecoratorFn, Meta, StoryObj } from "@storybook/react";
import {
  Priority1,
  Priority2,
  Priority3,
  Priority4,
  Priority5,
} from "@hitachivantara/uikit-react-icons";
import {
  HvDropdown,
  HvDropdownProps,
  HvDropdownStatus,
  HvGrid,
  HvListValue,
  theme,
} from "@hitachivantara/uikit-react-core";

const widthDecorator: DecoratorFn = (Story) => (
  <div style={{ minHeight: 120, width: 310 }}>{Story()}</div>
);

export default {
  title: "Components/Dropdown",
  component: HvDropdown,
  parameters: {
    eyes: {
      runBefore() {
        fireEvent.click(screen.getByRole("combobox"));
        return waitFor(() => screen.getByRole("listbox"));
      },
    },
  },
} as Meta<typeof HvDropdown>;

export const Main: StoryObj<HvDropdownProps> = {
  args: {
    multiSelect: true,
    showSearch: true,
    disabled: false,
    readOnly: false,
    required: false,
    defaultExpanded: true,
    notifyChangesOnFirstRender: false,
    hasTooltips: false,
    variableWidth: false,
    singleSelectionToggle: false,
    virtualized: false,
    status: "valid",
  },
  argTypes: {
    classes: { control: { disable: true } },
    label: { control: { disable: true } },
    popperProps: { control: { disable: true } },
  },
  parameters: {
    eyes: {
      runBefore() {},
    },
  },
  decorators: [
    widthDecorator,
    (Story) => <div style={{ minHeight: 400 }}>{Story()}</div>,
  ],
  render: (args) => (
    <HvDropdown
      label="Select values"
      values={[
        { label: "value 1" },
        { label: "value 2", selected: true },
        { label: "value 3" },
        { label: "value 4" },
      ]}
      {...args}
    />
  ),
};

export const Variants: StoryObj<HvDropdownProps> = {
  parameters: {
    docs: {
      description: {
        story: "Dropdown in their various form state variants.",
      },
    },
  },
  decorators: [
    (Story) => (
      <div
        className={css({
          display: "flex",
          flexFlow: "row wrap",
          gap: 16,
          "& > *": {
            width: 200,
          },
        })}
      >
        {Story()}
      </div>
    ),
  ],
  render: () => (
    <>
      <HvDropdown required label="Required" />
      <HvDropdown disabled label="Disabled" />
      <HvDropdown readOnly label="Read-only" />
      <HvDropdown status="invalid" label="Invalid" />
    </>
  ),
};

const PriorityIcon = ({
  Icon,
  label,
}: {
  Icon: React.ElementType;
  label: string;
}) => (
  <span
    className={css({
      lineHeight: "32px",
      display: "flex",
      alignItems: "center",
    })}
  >
    <Icon
      iconSize={{ width: 22, height: 22 }}
      className={css({ float: "left", margin: "5px 5px 5px 0" })}
    />
    <h3>{label}</h3>
  </span>
);

export const WithIcons: StoryObj<HvDropdownProps> = {
  parameters: {
    docs: {
      description: {
        story: "Single selection Dropdown with icons along with labels",
      },
    },
  },
  decorators: [widthDecorator],
  render: (args) => {
    const [values, setValues] = useState<HvListValue[]>([]);

    useEffect(() => {
      const icons = [Priority1, Priority2, Priority3, Priority4, Priority5];

      setValues(
        icons.map((Icon, i) => ({
          id: `p${i + 1}`,
          label: <PriorityIcon Icon={Icon} label={`Priority P${i + 1}`} />,
        }))
      );
    }, []);

    return (
      <HvDropdown aria-label="Dropdown With Icons" values={values} {...args} />
    );
  },
};

export const Empty: StoryObj<HvDropdownProps> = {
  parameters: {
    docs: {
      description: {
        story: "Dropdown with no values",
      },
    },
    eyes: { include: false },
  },
  decorators: [widthDecorator],
  render: () => <HvDropdown aria-label="Empty" />,
};

export const MultiSelection: StoryObj<HvDropdownProps> = {
  decorators: [widthDecorator],
  render: () => {
    return (
      <HvDropdown
        multiSelect
        showSearch
        label="Dropdown Title"
        values={[
          { label: "value 1" },
          { label: "value 2", selected: true },
          { label: "value 3" },
          { label: "value 4" },
        ]}
      />
    );
  },
};

export const SingleSelectionWithSearch: StoryObj<HvDropdownProps> = {
  parameters: {
    docs: {
      description: {
        story:
          "Single selection Dropdown with search and less than 10 elements",
      },
    },
  },
  decorators: [widthDecorator],
  render: () => (
    <HvDropdown
      aria-label="With search"
      showSearch
      values={[
        { label: "value 1" },
        { label: "value 2", selected: true },
        { label: "value 3" },
        { label: "value 4" },
      ]}
    />
  ),
};

export const ExternalErrorMessage: StoryObj<HvDropdownProps> = {
  parameters: {
    docs: {
      description: {
        story:
          "A form element can be invalid but render its error message elsewhere. For instance if a business rule error relates to the combination of two or more fields, or if we want to display all the form errors together in a summary section. The [aria-errormessage](https://w3c.github.io/aria/#aria-errormessage) property should reference another element that contains error message text. It can be used when controlling the validation status or when relying on the built-in validations, but the message text computation is reponsability of the app.",
      },
    },
    eyes: { include: false },
  },
  render: () => {
    const values1 = useMemo(
      () => [
        { label: "value 1" },
        { label: "value 2" },
        { label: "value 3" },
        { label: "value 4" },
      ],
      []
    );

    const values2 = useMemo(
      () => [
        { label: "value 1" },
        { label: "value 2" },
        { label: "value 3" },
        { label: "value 4" },
      ],
      []
    );

    const [deathValidationState, setDeathValidationState] = useState("invalid");

    const [birthErrorMessage, setBirthErrorMessage] = useState<string | null>(
      null
    );
    const [deathErrorMessage, setDeathErrorMessage] = useState(
      "Dropdown 2 is always invalid."
    );

    return (
      <HvGrid container>
        <HvGrid item xs={12} sm={6}>
          <HvDropdown
            label="Dropdown 1"
            multiSelect
            values={values1}
            required
            aria-errormessage="birth-error"
            onChange={(value) => {
              if ((value as HvListValue[]).length === 0) {
                setBirthErrorMessage(
                  "Select at least one value from dropdown 1."
                );
              } else {
                setBirthErrorMessage(null);
              }
            }}
          />
        </HvGrid>
        <HvGrid item xs={12} sm={6}>
          <HvDropdown
            label="Dropdown 2"
            multiSelect
            values={values2}
            required
            status={deathValidationState as HvDropdownStatus}
            aria-errormessage="death-error"
            onChange={(value) => {
              setDeathValidationState("invalid");

              if ((value as HvListValue[]).length === 0) {
                setDeathErrorMessage(
                  "Select at least one value from dropdown 2."
                );
              } else {
                setDeathErrorMessage(
                  `Dropdown 2 is always invalid, even with ${
                    (value as HvListValue[]).length
                  } items selected.`
                );
              }
            }}
          />
        </HvGrid>
        <HvGrid item xs={12}>
          <div
            style={{
              maxWidth: 600,
              paddingTop: "10px",
              paddingLeft: "20px",
              backgroundColor: theme.colors.negative_20,
              color: theme.colors.base_dark,
              height: "100%",
            }}
          >
            <h4>Form errors:</h4>
            <ul>
              {birthErrorMessage && (
                <li id="birth-error" aria-live="polite">
                  {birthErrorMessage}
                </li>
              )}
              {deathErrorMessage && (
                <li id="death-error" aria-live="polite">
                  {deathErrorMessage}
                </li>
              )}
            </ul>
          </div>
        </HvGrid>
      </HvGrid>
    );
  },
};

export const WithDefinedHeight: StoryObj<HvDropdownProps> = {
  parameters: {
    docs: {
      description: {
        story:
          "Dropdown's height can be configured using `height` (or `maxHeight`). Note: only validated in the single selection use-case.",
      },
    },
    eyes: { include: false },
  },
  decorators: [widthDecorator],
  render: () => {
    const values = [...Array(100)].map((_, i) => ({
      id: `${i}`,
      label: `value  ${i}`,
    }));

    return (
      <HvDropdown
        aria-label="With defined height"
        values={values}
        height={350}
        hasTooltips
        showSearch
      />
    );
  },
};

export const Virtualized: StoryObj<HvDropdownProps> = {
  parameters: {
    docs: {
      description: {
        story:
          "Experimental Dropdown with virtualized list, which handles performance in lists with a lot of options. Note: only validated in the single selection use-case.",
      },
    },
    eyes: { include: false },
  },
  decorators: [widthDecorator],
  render: () => {
    const values = [...Array(1500)].map((_, i) => ({
      id: `${i}`,
      label: `value  ${i}`,
    }));

    return (
      <HvDropdown
        aria-label="More than 1000 items"
        values={values}
        virtualized
        height={350}
        hasTooltips
        showSearch
      />
    );
  },
};
