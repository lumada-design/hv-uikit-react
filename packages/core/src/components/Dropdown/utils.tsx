import { HvListValue } from "@core/components";

/**
 * Filter selected elements.
 *
 * @param {Object} list - the list to filter
 * @returns {Array} - the selected elements
 */
const getSelected = (list) => list?.filter((elem) => elem.selected) || [];

/**
 * Checks if any element of the list is selected.
 *
 * @param list
 * @returns {boolean}
 */
const hasSelected = (list) => getSelected(list).length > 0;

/**
 * Gets the selection label according to selection.
 *
 * @param {Object} list - the list to filter the selected elements from
 * @param {Object} labels - the labels to extract the textual values for the label
 * @param {Boolean} multiSelect - if "true" the label will have a different format
 *
 * @returns {Object} - the selection label
 */
const getSelectionLabel = (
  labels,
  placeholder,
  multiSelect,
  list: HvListValue[] = []
) => {
  const { select } = labels;
  const selected = getSelected(list);

  if (select) return { selected: select };

  if (multiSelect) {
    return {
      selected: selected.length,
      total: list.length,
    };
  }
  return { selected: selected.length > 0 ? selected[0].label : placeholder };
};

export { getSelectionLabel, getSelected, hasSelected };
