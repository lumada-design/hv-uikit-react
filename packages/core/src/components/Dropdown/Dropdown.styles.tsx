import { theme } from "@hitachivantara/uikit-styles";

import { createClasses } from "@core/utils/classes";

export const { staticClasses, useClasses } = createClasses("HvDropdown", {
  root: {
    width: "100%",
    position: "relative",
    display: "inline-block",
  },
  labelContainer: { display: "flex", alignItems: "flex-start" },
  label: { paddingBottom: "6px", display: "block" },
  description: {},
  error: {},
  placeholder: {
    overflow: "hidden",
    textOverflow: "ellipsis",
    whiteSpace: "nowrap",
  },
  selectionDisabled: {
    lineHeight: theme.space.md,
    color: theme.dropdown.disabledColor,
  },
  dropdown: {
    width: "100%",
    borderRadius: theme.radii.base,

    "& $dropdownHeaderInvalid": {
      border: theme.dropdown.dropdownHeaderInvalidBorder,
      "&:hover": {
        border: theme.dropdown.dropdownHeaderInvalidBorder,
      },
    },
  },
  arrow: {},
  dropdownHeader: {},
  dropdownHeaderInvalid: {},
  dropdownHeaderOpen: {},
  dropdownListContainer: {},
  rootList: {},
});
