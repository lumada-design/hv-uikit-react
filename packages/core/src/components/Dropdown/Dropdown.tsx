import { forwardRef, useEffect, useRef, useState } from "react";

import { PopperProps } from "@mui/material/Popper";
import { useForkRef } from "@mui/material/utils";

import { useDefaultProps } from "@core/hooks/useDefaultProps";
import { setId } from "@core/utils/setId";
import { useLabels } from "@core/hooks/useLabels";
import { useUniqueId } from "@core/hooks/useUniqueId";
import { useControlled } from "@core/hooks/useControlled";
import { HvBaseProps } from "@core/types/generic";
import {
  HvBaseDropdown,
  HvBaseDropdownProps,
} from "@core/components/BaseDropdown";
import { HvListValue } from "@core/components/List";
import {
  isInvalid,
  HvInfoMessage,
  HvWarningText,
  HvFormElement,
  HvLabel,
} from "@core/components/Forms";
import { ExtractNames } from "@core/utils/classes";
import { HvTypography } from "@core/components/Typography";

import { getSelected, getSelectionLabel } from "./utils";
import { HvDropdownList, HvDropdownListProps } from "./List";
import { staticClasses, useClasses } from "./Dropdown.styles";
import { HvDropdownLabelsProps, HvDropdownStatus } from "./types";

export { staticClasses as dropdownClasses };

export type HvDropdownClasses = ExtractNames<typeof useClasses>;

export interface HvDropdownProps
  extends HvBaseProps<HTMLDivElement, "onChange"> {
  /**
   * A Jss Object used to override or extend the component styles applied.
   */
  classes?: HvDropdownClasses;
  /**
   * The form element name.
   */
  name?: string;
  /**
   * The label of the form element.
   *
   * The form element must be labeled for accessibility reasons.
   * If not provided, an aria-label or aria-labelledby must be provided instead.
   */
  label?: any;
  /**
   * Provide additional descriptive text for the form element.
   */
  description?: any;
  /**
   * The placeholder value when nothing is selected.
   */
  placeholder?: string;
  /**
   * Indicates that the form element is disabled.
   */
  disabled?: boolean;
  /**
   * Indicates that the form element is in read only mode.
   */
  readOnly?: boolean;
  /**
   * Indicates that user input is required on the form element.
   */
  required?: boolean;
  /**
   * The status of the form element.
   *
   * Valid is correct, invalid is incorrect and standBy means no validations have run.
   *
   * When uncontrolled and unspecified it will default to "standBy" and change to either "valid"
   * or "invalid" after any change to the state.
   */
  status?: HvDropdownStatus;
  /**
   * The error message to show when the validation status is "invalid".
   *
   * Defaults to "Required" when the status is uncontrolled and no `aria-errormessage` is provided.
   */
  statusMessage?: any;
  /**
   * Identifies the element that provides an error message for the dropdown.
   *
   * Will only be used when the validation status is invalid.
   */
  "aria-errormessage"?: string;
  /**
   * The callback fired when the value changes.
   */
  onChange?: (selected: HvListValue | HvListValue[] | undefined) => void;
  /**
   * The list to be rendered by the dropdown.
   */
  values?: HvListValue[];
  /**
   * If `true` the dropdown is multiSelect, if `false` the dropdown is single select.
   */
  multiSelect?: boolean;
  /**
   * If `true` the dropdown is rendered with a search bar, if `false` there won't be a search bar.
   */
  showSearch?: boolean;
  /**
   * If `true` the dropdown starts opened if `false` it starts closed.
   */
  expanded?: boolean;
  /**
   * When uncontrolled, defines the initial expanded state.
   */
  defaultExpanded?: boolean;
  /**
   * If 'true' the dropdown will notify on the first render.
   */
  notifyChangesOnFirstRender?: boolean;
  /**
   * An object containing all the labels for the dropdown.
   */
  labels?: HvDropdownLabelsProps;
  /**
   * If `true` the dropdown will show tooltips when user mouseenter text in list
   */
  hasTooltips?: boolean;
  /**
   * Disable the portal behavior.
   * The children stay within it's parent DOM hierarchy.
   */
  disablePortal?: boolean;
  /**
   * If `true` the dropdown width depends size of content if `false` the width depends on the header size.
   * Defaults to `false`.
   */
  variableWidth?: boolean;
  /**
   * If `true`, selection can be toggled when single selection.
   */
  singleSelectionToggle?: boolean;
  /**
   * Placement of the dropdown.
   */
  placement?: "left" | "right";
  /**
   * An object containing props to be wired to the popper component.
   */
  popperProps?: Partial<PopperProps>;
  /**
   * Callback called when the user cancels the changes.
   *
   * Called when the cancel button is used and when the user clicks outside the open container.
   *
   * @param {object} event The event source of the callback.
   */
  onCancel?: (event: Event) => void;
  /**
   * Callback called when dropdown changes the expanded state.
   *
   * @param {object} event The event source of the callback.
   * @param {boolean} open If the dropdown new state is open (`true`) or closed (`false`).
   */
  onToggle?: (event: Event, open: boolean) => void;
  /**
   * Callback called when the user clicks outside the open container.
   *
   * @param {object} event The event source of the callback.
   */
  onClickOutside?: (event: Event) => void;
  /**
   * @ignore
   */
  onFocus?: React.FocusEventHandler<any>;
  /**
   * @ignore
   */
  onBlur?: React.FocusEventHandler<any>;
  /**
   * Experimental. Height of the dropdown, in case you want to control it from a prop. Styles can also be used through dropdownListContainer class. Required in case virtualized is used
   */
  height?: number;
  /**
   * Experimental. Height of the dropdown, in case you want to control it from a prop. Styles can also be used through dropdownListContainer class. Required in case virtualized is used
   */
  maxHeight?: number;
  /**
   * Experimental. Uses dropdown in a virtualized form, where not all options are rendered initially. Good for use cases with a lot of options.
   */
  virtualized?: boolean;
  /**
   * Extra props passed to the dropdown.
   */
  baseDropdownProps?: Partial<HvBaseDropdownProps>;
  /**
   * Extra props passed to the list.
   */
  listProps?: HvDropdownListProps;
}

const DEFAULT_LABELS: HvDropdownLabelsProps = {
  select: undefined,
  selectAll: "All",
  cancelLabel: "Cancel",
  applyLabel: "Apply",
  searchPlaceholder: "Search",
  multiSelectionConjunction: "/",
};

/**
 * A dropdown list is a graphical control element, similar to a list box, that allows the user to choose one value from a list.
 */
export const HvDropdown = forwardRef<HTMLDivElement, HvDropdownProps>(
  (props, ref) => {
    const {
      classes: classesProp,
      className,

      id,
      name,

      required = false,
      disabled = false,
      readOnly = false,

      label,
      "aria-label": ariaLabel,
      "aria-labelledby": ariaLabelledBy,
      description,
      "aria-describedby": ariaDescribedBy,

      placeholder = "Select...",

      onChange,

      status,
      statusMessage,
      "aria-errormessage": ariaErrorMessage,

      onCancel,
      onToggle,
      onClickOutside,

      onFocus,
      onBlur,

      values,
      multiSelect = false,
      showSearch = false,
      expanded,
      defaultExpanded = false,
      notifyChangesOnFirstRender = false,
      labels: labelsProp,
      hasTooltips = false,
      disablePortal = false,
      singleSelectionToggle = true,
      placement,
      variableWidth = false,
      popperProps = {},
      height,
      maxHeight,
      virtualized = false,
      baseDropdownProps = {},
      listProps = {},
      ...others
    } = useDefaultProps("HvDropdown", props);

    const { classes, cx } = useClasses(classesProp);

    const labels = useLabels(DEFAULT_LABELS, labelsProp);

    const elementId = useUniqueId(id, "hvdropdown");

    const [validationState, setValidationState] = useControlled(
      status,
      "standBy"
    );

    const [validationMessage] = useControlled(statusMessage, "Required");

    const [isOpen, setIsOpen] = useControlled(
      expanded,
      Boolean(defaultExpanded)
    );
    const [selectionLabel, setSelectionLabel] = useState(
      getSelectionLabel(labels, placeholder, multiSelect, values)
    );

    const [internalValues, setInternalValues] = useState(values);

    // Hack - Keeping track of internal values for validation purposes since useState is async
    const internalValuesRef = useRef(values);

    useEffect(() => {
      setInternalValues(values);
      internalValuesRef.current = values;
    }, [values]);

    useEffect(() => {
      setSelectionLabel(
        getSelectionLabel(labels, placeholder, multiSelect, values)
      );
    }, [labels, multiSelect, placeholder, values]);

    if (import.meta.env.DEV && virtualized && !height) {
      // eslint-disable-next-line no-console
      console.error(
        "Dropdown/List in virtualized mode requires a height. Please define it."
      );
    }

    const dropdownHeaderRef = useRef<HTMLDivElement>();

    const {
      ref: refProp,
      dropdownHeaderRef: dropdownHeaderRefProp,
      ...otherBaseDropdownProps
    } = baseDropdownProps;
    const headerForkedRef = useForkRef(
      dropdownHeaderRefProp,
      dropdownHeaderRef
    );

    const dropdownForkedRef = useForkRef(ref, refProp);

    const handleToggle: HvBaseDropdownProps["onToggle"] = (event, open) => {
      onToggle?.(event, open);

      setIsOpen(open);

      if (!open) {
        // also run built-in validation when closing without changes
        // as the user "touched" the input
        setValidationState(() => {
          // this will only run if status is uncontrolled
          if (required) {
            const hasSelection =
              getSelected(internalValuesRef.current).length > 0;

            if (!hasSelection) {
              return "invalid";
            }
          }

          return "valid";
        });
      }
    };

    /** Applies the selected values to the state */
    const handleSelection: HvDropdownListProps["onChange"] = (
      listValues,
      commitChanges,
      toggle,
      notifyChanges = true
    ) => {
      const selected = getSelected(listValues);

      if (commitChanges) {
        setInternalValues(listValues);
        internalValuesRef.current = listValues;

        setSelectionLabel(
          getSelectionLabel(labels, placeholder, multiSelect, listValues)
        );

        setValidationState(() => {
          // this will only run if status is uncontrolled
          if (required && selected.length === 0) {
            return "invalid";
          }

          return "valid";
        });
      }
      if (notifyChanges) onChange?.(multiSelect ? selected : selected[0]);
      if (toggle) {
        handleToggle(undefined as any, false);

        // focus-ring won't be visible even if using the keyboard:
        // https://github.com/WICG/focus-visible/issues/88
        dropdownHeaderRef.current?.focus({ preventScroll: true });
      }
    };

    /**
     * Handles the `Cancel` action. Both single and ranged modes are handled here.
     */
    const handleCancel: HvDropdownListProps["onCancel"] = (evt) => {
      onCancel?.(evt as any);

      handleToggle(evt as any, false);

      // focus-ring won't be visible even if using the keyboard:
      // https://github.com/WICG/focus-visible/issues/88
      dropdownHeaderRef.current?.focus({ preventScroll: true });
    };

    const handleClickOutside: HvBaseDropdownProps["onClickOutside"] = (evt) => {
      onClickOutside?.(evt);
      onCancel?.(evt);
    };

    const setFocusToContent: HvBaseDropdownProps["onContainerCreation"] = (
      containerRef
    ) => {
      const inputs = containerRef?.getElementsByTagName("input");
      if (inputs && inputs.length > 0) {
        inputs[0].focus();
        return;
      }
      const listItems =
        containerRef != null
          ? [...containerRef.getElementsByTagName("li")]
          : [];
      listItems.every((listItem) => {
        if (listItem.tabIndex >= 0) {
          listItem.focus();
          return false;
        }
        return true;
      });
    };

    const buildHeaderLabel = () => {
      const hasSelection = getSelected(internalValues).length > 0;
      return labels?.select || !multiSelect ? (
        <HvTypography
          component="div"
          variant="body"
          className={cx(classes.placeholder, {
            [classes.selectionDisabled]: disabled,
            [classes.placeholderClosed]: !(isOpen || hasSelection),
          })}
        >
          {selectionLabel.selected}
        </HvTypography>
      ) : (
        <HvTypography
          component="div"
          className={cx(classes.placeholder, {
            [classes.selectionDisabled]: disabled,
          })}
          variant="body"
        >
          <b>{selectionLabel.selected}</b>
          {` ${labels?.multiSelectionConjunction} ${selectionLabel.total}`}
        </HvTypography>
      );
    };

    const hasLabel = label != null;
    const hasDescription = description != null;

    // the error message area will only be created if:
    // - an external element that provides an error message isn't identified via aria-errormessage AND
    //   - both status and statusMessage properties are being controlled OR
    //   - status is uncontrolled and required is true
    const canShowError =
      ariaErrorMessage == null &&
      ((status !== undefined && statusMessage !== undefined) ||
        (status === undefined && required));

    const isStateInvalid = isInvalid(validationState);

    let errorMessageId;
    if (isStateInvalid) {
      errorMessageId = canShowError
        ? setId(elementId, "error")
        : ariaErrorMessage;
    }

    return (
      <HvFormElement
        id={id}
        name={name}
        status={validationState}
        disabled={disabled}
        readOnly={readOnly}
        required={required}
        className={cx(
          classes.root,
          {
            [classes.disabled]: disabled,
          },
          className
        )}
        {...others}
      >
        {(hasLabel || hasDescription) && (
          <div className={classes.labelContainer}>
            {hasLabel && (
              <HvLabel
                id={setId(elementId, "label")}
                label={label}
                className={classes.label}
              />
            )}

            {hasDescription && (
              <HvInfoMessage
                id={setId(elementId, "description")}
                className={classes.description}
              >
                {description}
              </HvInfoMessage>
            )}
          </div>
        )}
        <HvBaseDropdown
          ref={dropdownForkedRef}
          id={setId(id, "dropdown")}
          classes={{
            root: cx(classes.dropdown, {
              [classes.readOnly]: readOnly,
            }),
            arrow: classes.arrow,
            header: cx(classes.dropdownHeader, {
              [classes.dropdownHeaderInvalid]: isStateInvalid,
            }),
            headerOpen: classes.dropdownHeaderOpen,
          }}
          expanded={isOpen}
          disabled={disabled}
          readOnly={readOnly}
          required={required}
          disablePortal={disablePortal}
          placement={placement}
          popperProps={popperProps}
          placeholder={buildHeaderLabel()}
          onToggle={handleToggle}
          onClickOutside={handleClickOutside}
          onContainerCreation={setFocusToContent}
          role="combobox"
          variableWidth={variableWidth}
          aria-label={ariaLabel}
          aria-labelledby={
            [label && setId(elementId, "label"), ariaLabelledBy]
              .join(" ")
              .trim() || undefined
          }
          aria-invalid={isStateInvalid ? true : undefined}
          aria-errormessage={errorMessageId}
          aria-describedby={
            [description && setId(elementId, "description"), ariaDescribedBy]
              .join(" ")
              .trim() || undefined
          }
          onFocus={onFocus}
          onBlur={onBlur}
          dropdownHeaderRef={headerForkedRef}
          {...otherBaseDropdownProps}
        >
          <HvDropdownList
            id={setId(elementId, "values")}
            classes={{
              rootList: classes.rootList,
              dropdownListContainer: classes.dropdownListContainer,
            }}
            values={internalValues}
            multiSelect={multiSelect}
            showSearch={showSearch}
            onChange={handleSelection}
            onCancel={handleCancel}
            labels={labels}
            notifyChangesOnFirstRender={notifyChangesOnFirstRender}
            hasTooltips={hasTooltips}
            singleSelectionToggle={singleSelectionToggle}
            aria-label={ariaLabel}
            aria-labelledby={hasLabel ? setId(elementId, "label") : undefined}
            height={height}
            maxHeight={maxHeight}
            virtualized={virtualized}
            {...listProps}
          />
        </HvBaseDropdown>
        {canShowError && (
          <HvWarningText
            id={setId(elementId, "error")}
            disableBorder
            className={classes.error}
          >
            {validationMessage}
          </HvWarningText>
        )}
      </HvFormElement>
    );
  }
);
