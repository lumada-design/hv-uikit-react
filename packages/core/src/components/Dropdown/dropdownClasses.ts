import { getClasses } from "@core/utils";

export interface HvDropdownClasses {
  root?: string;
  labelContainer?: string;
  label?: string;
  description?: string;
  error?: string;
  placeholder?: string;
  selectionDisabled?: string;
  dropdown?: string;
  arrow?: string;
  dropdownHeader?: string;
  dropdownHeaderInvalid?: string;
  dropdownHeaderOpen?: string;
  dropdownListContainer?: string;
  rootList?: string;
}

const classKeys: (keyof HvDropdownClasses)[] = [
  "root",
  "labelContainer",
  "label",
  "description",
  "error",
  "placeholder",
  "selectionDisabled",
  "dropdown",
  "arrow",
  "dropdownHeader",
  "dropdownHeaderInvalid",
  "dropdownHeaderOpen",
  "dropdownListContainer",
  "rootList",
];

const dropdownClasses = getClasses(classKeys, "HvDropdown");

export default dropdownClasses;
