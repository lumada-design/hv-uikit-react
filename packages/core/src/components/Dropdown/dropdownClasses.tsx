import { getClasses } from "@core/utils";

export type HvDropdownClasses = {
  root?: string;
  labelContainer?: string;
  label?: string;
  description?: string;
  error?: string;
  placeholder?: string;
  selectionDisabled?: string;
  dropdown?: string;
  arrow?: string;
  dropdownHeader?: string;
  dropdownHeaderInvalid?: string;
  dropdownHeaderOpen?: string;
  dropdownListContainer?: string;
  rootList?: string;
};

const classKeys: string[] = [
  "root",
  "labelContainer",
  "label",
  "description",
  "error",
  "placeholder",
  "selectionDisabled",
  "dropdown",
  "arrow",
  "dropdownHeader",
  "dropdownHeaderInvalid",
  "dropdownHeaderOpen",
  "dropdownListContainer",
  "rootList",
];

const dropdownClasses = getClasses<HvDropdownClasses>(classKeys, "HvDropdown");

export default dropdownClasses;
