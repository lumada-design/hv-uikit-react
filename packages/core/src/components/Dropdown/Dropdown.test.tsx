import { render, screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, expect, it } from "vitest";
import { HvDropdown } from "./Dropdown";

describe("Dropdown", () => {
  it("should open dropdown on click", async () => {
    render(
      <HvDropdown
        aria-label="Main sample"
        multiSelect
        showSearch
        values={[
          { label: "value 1" },
          { label: "value 2", selected: true },
          { label: "value 3" },
          { label: "value 4" },
        ]}
      />
    );

    const dropdown = screen.getByRole("combobox");

    expect(dropdown).toHaveAttribute("aria-expanded", "false");

    await userEvent.click(dropdown);

    expect(dropdown).toHaveAttribute("aria-expanded", "true");
  });

  it("should be disabled", async () => {
    const { container } = render(
      <HvDropdown
        aria-label="Main sample"
        multiSelect
        showSearch
        disabled
        values={[
          { label: "value 1" },
          { label: "value 2", selected: true },
          { label: "value 3" },
          { label: "value 4" },
        ]}
      />
    );
    expect(
      container.getElementsByClassName("HvBaseDropdown-rootDisabled")
    ).toHaveLength(1);

    const DropdownHeader = screen.getByRole("combobox");

    expect(DropdownHeader).toHaveAttribute("aria-expanded", "false");

    expect(DropdownHeader).toHaveAttribute("aria-expanded", "false");
  });

  it("should be invalid", async () => {
    render(
      <HvDropdown
        aria-label="Main sample"
        multiSelect
        showSearch
        status="invalid"
        values={[
          { label: "value 1" },
          { label: "value 2", selected: true },
          { label: "value 3" },
          { label: "value 4" },
        ]}
      />
    );

    const DropdownHeader = screen.getByRole("combobox");

    expect(DropdownHeader).toHaveAttribute("aria-invalid", "true");
  });

  it("should select one", async () => {
    render(
      <HvDropdown
        aria-label="Main sample"
        multiSelect
        showSearch
        values={[
          { label: "value 1" },
          { label: "value 2" },
          { label: "value 3" },
          { label: "value 4" },
        ]}
      />
    );

    const DropdownHeader = screen.getByRole("combobox");

    expect(DropdownHeader).toHaveAttribute("aria-expanded", "false");

    await userEvent.click(DropdownHeader);

    expect(DropdownHeader).toHaveAttribute("aria-expanded", "true");

    const options = screen.getAllByRole("listitem");

    expect(options).toHaveLength(4);
    expect(within(options[0]).getByRole("checkbox")).not.toBeChecked();

    await userEvent.click(options[0]);
    expect(within(options[0]).getByRole("checkbox")).toBeChecked();
  });

  it("read Only", async () => {
    render(
      <HvDropdown
        aria-label="Main sample"
        multiSelect
        showSearch
        readOnly
        values={[
          { label: "value 1" },
          { label: "value 2" },
          { label: "value 3" },
          { label: "value 4" },
        ]}
      />
    );

    const DropdownHeader = screen.getByRole("combobox");

    expect(DropdownHeader).toHaveAttribute("aria-expanded", "false");

    await userEvent.click(DropdownHeader);

    expect(DropdownHeader).toHaveAttribute("aria-expanded", "false");
  });
});
