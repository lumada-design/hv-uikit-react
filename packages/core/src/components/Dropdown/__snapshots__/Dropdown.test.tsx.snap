// Vitest Snapshot v1

exports[`Dropdown > should render correctly 1`] = `
<div>
  <div
    style="width: 310px;"
  >
    <div
      class="HvDropdown-root css-13v24hy-StyledHvFormElement etta2py4 HvFormElement-root"
    >
      <div
        class="HvBaseDropdown-root HvDropdown-dropdown css-1nukton-StyledRoot eapzthe9"
      >
        <div
          aria-expanded="false"
          aria-label="Main sample"
          class="etta2py3 HvBaseDropdown-anchor css-151d1mt-StyledAnchor-StyledDropdown eapzthe8"
          role="combobox"
          tabindex="-1"
        >
          <div
            aria-label="Main sample"
            class="HvBaseDropdown-header HvDropdown-dropdownHeader css-z15in3-StyledHeaderRoot eapzthe7"
            role="textbox"
            tabindex="0"
          >
            <div
              class="HvBaseDropdown-selection css-1n3tnd9-StyledSelection eapzthe6"
            >
              <div
                class="HvDropdown-placeholder etta2py0 css-4kh4g2-getStyledComponent-StyledTypography e1tnpalo0"
              >
                <b>
                  1
                </b>
                 / 4
              </div>
            </div>
            <div
              class="HvBaseDropdown-arrow HvDropdown-arrow css-tgcv4t-StyledDropDownXS eapzthe3 css-du18qm"
              name="DropDownXS"
            >
              <svg
                focusable="false"
                height="12"
                viewBox="0 0 12 12"
                width="12"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  class="color0"
                  d="M5.99994 8.65 1.4 4.05l.7-.7L6 7.25l3.9-3.9.7.7z"
                  fill="var(--colors-acce1)"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
